jax/api.py

```diff
@@ -2081,20 +2085,27 @@ def device_put_sharded(x: Sequence[Any], devices: Sequence[xc.Device]):
   - device_put
   - device_put_replicated
   """
-  if not isinstance(x, Sequence):
-    raise ValueError(f"x must be a sequence; got {type(x)}")
-  # TODO(jakevdp): provide a default for devices that considers both local devices and pods
-  assert len(x) == len(devices), f"len(x) = {len(x)} must equal len(devices) = {len(devices)}."
+  # TODO(jakevdp): provide a default for devices that considers both local
+  # devices and pods
+  if not isinstance(shards, Sequence):
+    raise ValueError("device_put_sharded `shards` input must be a sequence; "
+                     f"got {type(shards)}")
+  if not len(shards) == len(devices):
+    raise ValueError(f"len(shards) = {len(shards)} must equal "
+                     f"len(devices) = {len(devices)}.")
+
   def _device_put_sharded(*xs) -> pxla.ShardedDeviceArray:
     avals = [core.raise_to_shaped(core.get_aval(x)) for x in xs]
-    # We cannot check aval equality directly, because it may fail for ConcreteArray.
-    assert all(aval.shape == avals[0].shape and aval.dtype == avals[0].dtype for aval in avals),\
-      f"abstract values not compatible: {avals}"
-    x_aval = core.raise_to_shaped(avals[0])
-    aval = ShapedArray((len(devices),) + x_aval.shape, x_aval.dtype)
-    buffers = list(it.chain.from_iterable(xla.device_put(x, d) for x, d in zip(xs, devices)))
-    return pxla.ShardedDeviceArray(aval, buffers)
-  return tree_multimap(_device_put_sharded, *x)
+    if not all(a1 == a2 for a1, a2 in zip(avals[:-1], avals[1:])):
```

Comment:
Do we need to worry about the avals being concrete arrays here? If they are, the equality check will check value, rather than just shape/dtype.

Reply:
Good thing to check! I don't think so, because of the `core.raise_to_shaped`. (I wondered if perhaps the concern about ConcreteDeviceArrays was left over from some point when `raise_to_shaped` wasn't called.)

Reply:
Ah, right... that makes sense.